### PR TITLE
[TT-645] Use standard in handling of JWK

### LINKED ISSUE #2303: Validate JWT using PEM rather than x509 certificate
**Do you want to request a *feature* or report a *bug*?**

Feature

**What is the current behavior?**

Receive the following error when setting up a Tyk JWT authorizer with Okta as an authorization server:

```
curl 'https://$tyk-api/$path-with-jwt-authorizer' -H 'User-Agent: Mozilla/5.0 (Macintosh; Intel Mac OS X 10.14; rv:67.0) Gecko/20100101 Firefox/67.0' -H 'Accept: */*' -H 'Accept-Language: en-US,en;q=0.5' --compressed -H 'Referer: http://localhost:3000/protected' -H 'Authorization: Bearer $okta-access-token' -H 'Origin: http://localhost:3000' -H 'DNT: 1' -H 'Connection: keep-alive'
```

```json
{
    "error": "Key not authorized:no certificates in JWK"
}
```

This is due to Tyk expecting an `x5c` parameter: https://github.com/TykTechnologies/tyk/blob/f69055a931dc50abf3ef884fa9cb7277ae667e4c/gateway/mw_jwt.go#L94-L104 and not using `n` and `e` to generate a PEM if they are present.

**What is the expected behavior?**

Tyk verifies the JWT using the PEM modulo and exponent in Okta's JWT keyset.

Okta do not expose their X509 certificate in their JWKS endpoint anymore (https://stackoverflow.com/questions/38727528/is-okta-oidc-documentation-outdated/38947218#38947218)).

Example keyset response parameters (values omitted):

```
{
    "keys": [
        {
            "alg": "RS256",
            "e": ...,
            "kid": ...,
            "kty": "RSA",
            "n": ...,
            "use": "sig"
        }
    ]
}
```

**Which versions of Tyk affected by this issue? Did this work in previous versions of Tyk?**

All versions (`<=v2.8.0` at the time of writing).

### PR DESCRIPTION
**Pull request by gernest**
## Description

This adds support for using JWK with missing certificate field. Currently we
were expecting jwk with the following format.

```json
{
  "keys": [
    {
      "alg": "RS256",
      "kty": "RSA",
      "use": "sig",
      "x5c": [
        "Ci0tLS0tQkVHSU4gUFVCTElDIEtFWS0tLS0tCk1JSUJJakFOQmdrcWhraUc5dzBCQVFFRkFBT0NBUThBTUlJQkNnS0NBUUVBeXFaNHJ3S0Y4cUNFeFM3a3BZNGMKbkphLzM3Rk1rSk5rYWxaM091c2xMQjBvUkw4VDRjOTRrZEY0YWVOelNGa1NlMm45OUlCSTZTc2w3OXZiZk1aYgordDA2TDBROTRrKy9QMzd4NysvUkpaaWZmNHkxVkdqcm5ybk1JMml1OWw0aUJCUll6Tm1HNmVibHJvRU1NV2xnCms1dHlzSGd4QjU5Q1NOSWNEOWdxazFoeDRuL0ZnT212S3NmUWdXSE5sUFNEVFJjV0dXR2hCMi9YZ05WWUcycE8KbFF4QVBxTGhCSGVxR1RYQmJQZkdGOWNIeml4cHNQcjZHdGJ6UHdoc1EvOGJQeG9KN2hkZm4rcnp6dGtzM2Q2KwpIV1VSY3lOVExSZTBtalhqamVlOVo2K2daK0grZlM0cG5QOXRxVDdJZ1U2ZVBVV1Rwam9pUHRMZXhnc0FhL2N0CmpRSURBUUFCCi0tLS0tRU5EIFBVQkxJQyBLRVktLS0tLQo="
      ],
      "n": "xofiG8gsnv9-I_g-5OWTLhaZtgAGq1QEsBCPK9lmLqhuonHe8lT-nK1DM49f6J9QgaOjZ3DB50QkhBysnIFNcXFyzaYIPMoccvuHLPgdBawX4WYKm5gficD0WB0XnTt4sqTI5usFpuop9vvW44BwVGhRqMT7c11gA8TSWMBxDI4A5ARc4MuQtfm64oN-JQodSztArwb9wcmH8WrBvSUkR4pyi9MT8W27gqJ2e2Xn8jgGnswNQWOyCTN84PawOYaN-2ORHeIea1g-URln1bofcHN73vZCIrVbE6iA2D7Ybh22AVrCfunekEDEe2GZfLZLejiZiBWG7enJhcrQIzAQGw",
      "e": "AQAB",
      "kid": "12345",
      "x5t": "12345"
    }
  ]
}
```

Where we used first entry on `x5c` which can either be a valid rsa key or  a `x509` certificate to derive the public key needed for verification of jwt tokens.

However services like okta they don't provide `x5c` instead you get the key from the `n` and `e` field.JWK response from okta looks like this.

```json
{
  "keys": [
    {
      "use": "sig",
      "kty": "RSA",
      "kid": "12345",
      "alg": "RS256",
      "n": "yqZ4rwKF8qCExS7kpY4cnJa_37FMkJNkalZ3OuslLB0oRL8T4c94kdF4aeNzSFkSe2n99IBI6Ssl79vbfMZb-t06L0Q94k-_P37x7-_RJZiff4y1VGjrnrnMI2iu9l4iBBRYzNmG6eblroEMMWlgk5tysHgxB59CSNIcD9gqk1hx4n_FgOmvKsfQgWHNlPSDTRcWGWGhB2_XgNVYG2pOlQxAPqLhBHeqGTXBbPfGF9cHzixpsPr6GtbzPwhsQ_8bPxoJ7hdfn-rzztks3d6-HWURcyNTLRe0mjXjjee9Z6-gZ-H-fS4pnP9tqT7IgU6ePUWTpjoiPtLexgsAa_ctjQ",
      "e": "AQAB"
    }
  ]
}
```

In this PR we are using the library `github.com/square/go-jose` for parsing JWK with missing `s5c`. To maintain backward compatibility we parse the JWK twice with our current implementation and with `go-jose` because tyk doesn't respect `x5c` to be a valid certificate , tyk works with rsa keys inside `x5c` which is not standard so we  leave certificate handling routines to preserve backward compatibility. For the case of missing `x5c` we use `go-jose` parsed keys.
## Related Issue
Based on #3206
fixes #2303

## How This Has Been Tested
`go test ./gateway`

- [ ] TODO with okta?

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Refactoring or add test (improvements in base code or adds test coverage to functionality)

## Checklist

- [ ] Make sure you are requesting to **pull a topic/feature/bugfix branch** (right side). If pulling from your own
      fork, don't request your `master`!
- [ ] Make sure you are making a pull request against the **`master` branch** (left side). Also, you should start
      *your branch* off *our latest `master`*.
- [ ] My change requires a change to the documentation.
  - [ ] If you've changed APIs, describe what needs to be updated in the documentation.
  - [ ] If new config option added, ensure that it can be set via ENV variable
- [ ] I have updated the documentation accordingly.
- [ ] Modules and vendor dependencies have been updated; run `go mod tidy && go mod vendor`
- [ ] When updating library version must provide reason/explanation for this update.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.
- [ ] Check your code additions will not fail linting checks:
  - [ ] `go fmt -s`
  - [ ] `go vet`

### CODE REVIEW

**Comment by gernest** [comment]
Gentle ping @letzya

**Comment by bmonteiro** [comment]
tested looks good

**Comment by bmonteiro** [comment]
Just one additional comment, if we use e and n why keeping x5c :) let's remove it because it is optimal from rfc and n is enough. Makes sense ? @gernest @letzya

**Comment by gernest** [comment]
@bmonteiro we have to maintain backward compatibility.

**Comment by bmonteiro** [comment]
There is no need "n" is a must and x5c optional. So it will work

**Comment by letzya** [comment]
@bmonteiro don't think it's not going to work since we need to keep backward compatibility. If for some reason a user only used x5c, his api now will stop working.

**Comment by gernest** [comment]
> There is no need "n" is a must and x5c optional. So it will work. It will be impossible having an IDP not sending "n" ;)

Prior to this change, tyk actually required `x5c` to work so `x5c` was  a must, that's why this bug manifested. Now we need to keep that behaviour and handle  the edge case of when `x5c` is not there.

**Comment by bmonteiro** [comment]
Current implementation when you consider x5c a must is actually not correct based on the RFC ;)

**Comment by bmonteiro** [comment]
> @bmonteiro don't think it's not going to work since we need to keep backward compatibility. If for some reason a user only used x5c, his api now will stop working. 

The user will also have from the IdP the "n", right ?

**Comment by gernest** [comment]
> Current implementation when you consider x5c a must is actually not correct based on the RFC ;)
I think thats what this PR is trying to address without breaking users who were relying on `x5c`.

**Comment by bmonteiro** [comment]
> > Current implementation when you consider x5c a must is actually not correct based on the RFC ;)
> 
> 
> 
> I think thats what this PR is trying to address without breaking users who were relying on `x5c`.

Ok guys, not to drag too much the chat. Long story short the IdP will pass the public key via "N" and "x5c" is optional and some do pass it others no. So if you remove x5c completely even old customers assuming they are using x5c will start using silently the "N"

**Comment by bmonteiro** [comment]
> @bmonteiro I'm not sure. potentially if you scrap a JWKs with just x5c and alg and kid it should work, so clients that don't follow the standard as-is are still **appreciated** and **close to our hearts** and should still be able to upgrade :)

The fundamental question is: There are out-there IdPs issuing or exposing non standard JWKs ?  What I mean is: if all is based on RFCs such scenario will never happen. Do you have any Idp doing that? Not passing 'N' that is the public key?

**Comment by buger** [comment]
@gernest let's make it use x5c only if N is not presented.

File: gateway/testutil.go [hunk 1]
```diff
@@ -463,16 +467,30 @@ func restDataSourceHandler(w http.ResponseWriter, r *http.Request) {
...
+            "kty": "RSA",
+            "kid": "12345",
+            "alg": "RS256",
+            "n": "yqZ4rwKF8qCExS7kpY4cnJa_37FMkJNkalZ3OuslLB0oRL8T4c94kdF4aeNzSFkSe2n99IBI6Ssl79vbfMZb-t06L0Q94k-_P37x7-_RJZiff4y1VGjrnrnMI2iu9l4iBBRYzNmG6eblroEMMWlgk5tysHgxB59CSNIcD9gqk1hx4n_FgOmvKsfQgWHNlPSDTRcWGWGhB2_XgNVYG2pOlQxAPqLhBHeqGTXBbPfGF9cHzixpsPr6GtbzPwhsQ_8bPxoJ7hdfn-rzztks3d6-HWURcyNTLRe0mjXjjee9Z6-gZ-H-fS4pnP9tqT7IgU6ePUWTpjoiPtLexgsAa_ctjQ",
+            "e": "AQAB"
+        }
+    ]
+}`
+
+const jwkTestJsonLegacy = `{
```
Comment (bmonteiro) [inline gateway/testutil.go:482]: I don't believe this test is valid because N is always present. I have never seen an IDP not sending N.

**Comment by buger** [comment]
@bmonteiro re-read the standard one more time.

`n` and `e` presented only for RSA keys. If you have ECDSA it will be another set of params. 
So having `x5c` as a fallback still makes sense. 

Additionally, even if both x5c and `n`/`e` are presented, x5c[0] cert public key match key from `n`,`e`.

**Comment by bmonteiro** [comment]
> ECDSA

@buger sure, I can read but it is also available for you here a good written example: https://tools.ietf.org/html/draft-jones-json-web-key-00

Indeed ECDSA uses curve and x and y (4.2.1.  JWK Key Object Members for ECDSA Keys). 

Having a hard time to understand and see where the x5c is but I'm sure it is me not getting it.

RSA also (4.2.2.  JWK Key Object Members for RSA Keys)

They explain it pretty nice with MUST be present fields but still I cannot see x5c there.

Long story short, no need to continue this because obviously you guys got it right.

**Comment by letzya** [comment]
@gernest is this still on our plan to be merged?

**Comment by gernest** [comment]
@letzya I don't think so.  It's not on jira. You can help create a jira ticket and link this PR to it, maybe it can be squeezed into any of the next release. Its pretty much done by now.

I will try to look at this on monday.

**Review by matiasinsaurralde — APPROVED** [review]

**Comment by edwardsmatt** [comment]
Hi there,
FWIW I have successfully tried this branch with AWS Cognito User Pool well-known keys.

Is there anything I can provide specifically to support this PR?

If not, are there any indications of when this may be merged/released?

Thanks very much

**Comment by matiasinsaurralde** [comment]
Conflicts solved but we still need additional reviews for merging.

**Review by edwardsmatt — APPROVED** [review]

**Comment by matiasinsaurralde** [comment]
Merged into `release-3-lts` (4bcf6dccdc7fbbc25640f7792758e1cf99ffc8da) and `release-3.1` (d24b90eb7ae72cfa626d8f6b0f1f2e147aec5afd).

**Review by tbuchaillot — APPROVED** [review]
